Active Record Associations Documentation

Ridhwana opened the pull request:
### Motivation / Background

Update the Active Record Association Documentation.

### Additional information

- [x] The images used to show the relationships of the different types of associations have both the table declarations, as well as a code sample, but the code sample is also in the guide. We could probably simplify the images by cutting off the code samples from them and leaving them in the guide code.
- [x] belongs_to section mentions bi-directional association and links to it within the same guide, but doesn't do the same for has_one and has_many, mentioned right after in the same paragraph.
- [x] There's a note about using bigint :supplier_id in the section between belongs_to and has_one, but up until that point it's always been using belongs_to in migrations... it's probably something that doesn't need to be in this area of the guide.
- [x] The has_many :through vs has_and_belongs_to_many section ends with a composite keys mention, we could link to the new guide there.
- [x] Associations between models with composite PKs may be simplified.
- [x] One example in bi-directional associations mentions Book belongs_to :writer, but then uses book.author a few lines below, which is incorrect.
- [x] STI section could be expanded, it's very useful in certain situations.
- [x] Add more info and make it clear

### Checklist

Before submitting the PR make sure the following are checked:

* [ ] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [ ] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [ ] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

Ridhwana committed:
update the wording of the intro section

Ridhwana committed:
group the sections differently

Ridhwana committed:
update some of the headers to be less repetitive

Ridhwana committed:
intro for associations

Ridhwana committed:
move content around to the introduction

Ridhwana committed:
newline

Ridhwana committed:
has_one association updates

Ridhwana committed:
has_many association updates

Ridhwana committed:
swap the code sample and diagram

Ridhwana committed:
has_many through

Ridhwana committed:
has_many :through

Ridhwana committed:
has_one :through

Ridhwana committed:
habtm

Ridhwana committed:
belongs_to vs has_one

Ridhwana committed:
has_many :through vs has_and_belongs_to_many

Ridhwana committed:
update has_many :through docs

Ridhwana committed:
polymorphic guide

Ridhwana committed:
more updates to the polymorphic guide

Ridhwana committed:
more updates to the polymorphic guide

Ridhwana committed:
composite primary keys

Ridhwana committed:
composite primary keys updates

Ridhwana committed:
self joins

Ridhwana committed:
add some more detail to the tips sections

Ridhwana committed:
bidirectional relationships

Ridhwana committed:
update the association reference

Ridhwana committed:
update the association references

Ridhwana committed:
Association Callbacks

Ridhwana committed:
STI's

Ridhwana committed:
STI's updates

Ridhwana committed:
delegated types

Ridhwana committed:
update the links to avoid using 'here'

Ridhwana committed:
polish up the docs

Ridhwana committed:
polish up the docs

p8 commented on guides/source/association_basics.md:
````diff
@@ -30,13 +44,16 @@ class Book < ApplicationRecord
+value when creating the book.
 
 ```ruby
 @book = Book.create(published_at: Time.now, author_id: @author.id)
````
Maybe remove the `published_at`, as it's not relevant to the example?
```suggestion
@book = Book.create(author_id: @author.id)
```

Ah, nevermind. It makes sense with the later examples that pass attributes to the association.
Maybe pass the `author_id` first though?
```suggestion
@book = Book.create(author_id: @author.id, published_at: Time.now)
```

p8 commented on guides/source/association_basics.md:
```diff
@@ -7,20 +7,34 @@ This guide covers the association features of Active Record.
+
+Without associations, creating and deleting books for that author would
+require a tedious and manual process. Here's what that would look like:
 
```
It's a bit unclear how `author_id` is defined on Book. I think it would help if we show the migrations for Authors and Books:
```suggestion

```ruby
class CreateAuthors < ActiveRecord::Migration[7.2]
  def change
    create_table :authors do |t|
      t.string :name
      t.timestamps
    end
    
    create_table :books do |t|
      t.belongs_to :author
      t.datetime :published_at
      t.timestamps
    end
  end
end

```

p8 commented on guides/source/association_basics.md:
```diff
@@ -107,7 +150,12 @@ end
+plural form, like `belongs_to :authors` in the `Book` model, and try to create a
+book with `Book.create(authors: @author)`, Rails will give you an "uninitialized
+constant Book::Authors" error. This happens because Rails automatically infers
+the class name from the association name. If the association name is pluralized,
```
I think Rails just creates a constant from the association name. This currently might suggest Rails does something special for pluralized names:
```suggestion
the class name from the association name. If the association name is `:authors`,
```

p8 commented on guides/source/association_basics.md:
```diff
@@ -131,7 +179,16 @@ end
 To setup a [bi-directional association](#bi-directional-associations) - use `belongs_to` in combination with a `has_one` or `has_many` on the other model, in this case the Author model.
 
-`belongs_to` does not ensure reference consistency if `optional` is set to true, so depending on the use case, you might also need to add a database-level foreign key constraint on the reference column, like this:
+If `optional` is set to true in the model, then `belongs_to` does not guarantee reference consistency. This means that the foreign key in one table might not reliably point to a valid primary key in the referenced table.
```
We don't mention belongs_to "guarantees reference consistency" before mentioning the `optional` option.
Maybe we can prepend that explanation:
```suggestion
By default `belongs_to` validates the presence of the association to guarantee reference consistency.

If `optional` is set to true in the model, then `belongs_to` does not guarantee reference consistency. This means that the foreign key in one table might not reliably point to a valid primary key in the referenced table.
```

p8 commented on guides/source/association_basics.md:
```diff
@@ -383,7 +506,7 @@ end
 ![has_and_belongs_to_many Association Diagram](images/association_basics/habtm.png)
 
-The corresponding migration might look like this:
+Even though a `has_and_belongs_to_many` does not require an intervening model, it does require a separate table to establish the many-to-many relationship between the two models involved. This intervening table serves to store the related data, mapping the associations between instances of the two models. UIt does not necessarily need a primary key since its purpose is solely to manage the relati ...
```
Small typo, but maybe change it to `The table` for clarity as well?
```suggestion
Even though a `has_and_belongs_to_many` does not require an intervening model, it does require a separate table to establish the many-to-many relationship between the two models involved. This intervening table serves to store the related data, mapping the associations between instances of the two models. The table does not necessarily need a primary key since its purpose is solely to manage the relationship between the associated records. The corresponding migration might look like this:
```

OughtPuts replied:
Typo - should be 'It does not necessarily need a primary key'

p8 commented on guides/source/association_basics.md:
```diff
@@ -406,11 +529,39 @@ class CreateAssembliesAndParts < ActiveRecord::Migration[7.2]
+not need a primary key of its own. `t.belongs_to :assembly` and `t.belongs_to
+:part` creates foreign keys linking the join table to the `assemblies` and
+`parts` tables respectively, ensuring [referential
+integrity](https://en.wikipedia.org/wiki/Referential_integrity).
```
I'm wondering if it would make sense to make these comments in the migration instead?
It would require less focus switching between this paragraph an the migration.

p8 commented on guides/source/association_basics.md:
```diff
@@ -406,11 +529,39 @@ class CreateAssembliesAndParts < ActiveRecord::Migration[7.2]
+WARNING: The use of extra attributes on the join table in a `has_and_belongs_to_many` association is deprecated. If you require this sort of complex behavior on the table that joins two models in a many-to-many relationship, you should use a `has_many :through` association instead of `has_and_belongs_to_many`.
+
+Choosing an Association
+-------------------------
```
```suggestion
-----------------------
```

p8 commented on guides/source/association_basics.md:
```diff
@@ -406,11 +529,39 @@ class CreateAssembliesAndParts < ActiveRecord::Migration[7.2]
+
+### `belongs_to` vs `has_one`
 
 If you want to set up a one-to-one relationship between two models, you'll need to add `belongs_to` to one, and `has_one` to the other. How do you know which is which?
```
Adding both associations is only needed if you want to define a bi-directional relationship. It's perfectly fine to start with defining an association on one side only.
Maybe something like:
```suggestion
If you want to set up a one-to-one relationship between two models, you can choose between a `belongs_to` and a `has_one` association. How do you know which one to choose?
```

Ridhwana replied:
Good catch, thank you!

p8 commented on guides/source/association_basics.md:
```diff
@@ -478,16 +619,37 @@ class Part < ApplicationRecord
+- The association is simple and does not require additional attributes or behaviors on the join table.
+- You do not need validations, callbacks, or extra methods on the join table.
+
+Other Associations
+-------------------------
```
```suggestion
Other Associations
-------------------
```

Or maybe we can rename this section to be more descriptive?
For example:
```suggestion
Advanced Associations
--------------------
```

Or maybe `Special Associations`?

p8 replied:
Maybe some of these sub sections can be upgraded to top sections instead?
It seems "Polymorphic Associations", "Self-Joins", "STI" and "Delegated Types" are somewhat on the same abstraction level?

Ridhwana replied:
> Maybe some of these sub-sections can be upgraded to top sections instead?
It seems "Polymorphic Associations", "Self-Joins", "STI" and "Delegated Types" are somewhat on the same abstraction level?

I would have liked to group them but the sub headings for STI and Delegated types are useful, and we lose the index under those headings once we move them under another subheading like Advanced Associations. 

Maybe another option would be to bring them closer up to the top, after "Advanced Associations", so that they'll be sections 5 and 6 and everything else will move further down. 

What are your thoughts?

p8 replied:
Moving them after "Advanced Assciations" sounds great!

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -7,20 +7,34 @@ This guide covers the association features of Active Record.
 -----------------
 
-In Rails, an _association_ is a connection between two Active Record models. Why do we need associations between models? Because they make common operations simpler and easier in your code.
+_Associations_ in Rails are special macro style calls that make it easy to
```
Link here providing the option for more context on macro?

Ridhwana replied:
Good point! I added an info block under:

INFO: A macro-style call is a method that generates or modifies other methods at
runtime, allowing for concise and expressive declarations of functionality, such
as defining model associations in Rails. For example, `has_many :comments`.

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -7,20 +7,34 @@ This guide covers the association features of Active Record.
+_Associations_ in Rails are special macro style calls that make it easy to
+connect different models. You can use them to tell Rails how your models relate
+to each other, which helps you manage your data more effectively, and makes
+common operations simpler and easier in your code.
```
Suggested change: 'easier to read' in your code - would suggest that readability is a big part of this.

Ridhwana replied:
✅

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -131,7 +179,16 @@ end
 To setup a [bi-directional association](#bi-directional-associations) - use `belongs_to` in combination with a `has_one` or `has_many` on the other model, in this case the Author model.
 
-`belongs_to` does not ensure reference consistency if `optional` is set to true, so depending on the use case, you might also need to add a database-level foreign key constraint on the reference column, like this:
+If `optional` is set to true in the model, then `belongs_to` does not guarantee reference consistency. This means that the foreign key in one table might not reliably point to a valid primary key in the referenced table.
```
Might be worth putting an "NB" at the start of this paragraph to signify there's a gotcha here. I was confused for a while about this, because this almost reads like the primary purpose of setting optional: true is not to guarantee reference consistency...

Ridhwana replied:
Sounds good, I've turned this into a note.

OughtPuts commented on guides/source/association_basics.md:
````diff
@@ -152,7 +211,7 @@ class Supplier < ApplicationRecord
 ```
 
-The main difference from `belongs_to` is that the link column `supplier_id` is located in the other table:
+The main difference from `belongs_to` is that the link column `supplier_id` is located in the other table, not the table where the `has_one` is declared.
````
suggested change: 'the link column (in this case, `supplier_id`)'

Ridhwana replied:
✅

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -175,9 +234,12 @@ class CreateSuppliers < ActiveRecord::Migration[7.2]
+Depending on the use case, you might also need to create a unique index and/or a
+foreign key constraint on the supplier column for the accounts table. The unique
+index ensures that each supplier is associated with only one account and allows
+you to query in a performant manner, while the foreign key constraint ensures
```
Might need a reword so that a beginner understands what 'performant' is in the context of queries. 'Allows you to query in an efficient manner?' maybe? Or 'allows you to query with the the fewest possible database hits' or something like that.

Ridhwana replied:
✅

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -478,16 +619,37 @@ class Part < ApplicationRecord
 
-You should use `has_many :through` if you need validations, callbacks, or extra attributes on the join model.
+- You need to add extra attributes or methods to the join table.
+- You require validations or callbacks on the join model.
```
Worth linking to validations and callbacks here in case beginners are unclear what this means?

Ridhwana replied:
✅

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -524,7 +691,9 @@ class CreatePictures < ActiveRecord::Migration[7.2]
-This migration can be simplified by using the `t.references` form:
+In our example, `imageable_id` could be the ID of either an `Employee` or a `Product`, and `imageable_type` is the name of the associated model's class, so either `Employee` or `Product`.
+
+While craeting the polymorphic association manually is acceptable, it is instead recommeded to use `t.references`and specify `polymorphic: true` so that Rails knows that the association is polymorphic, and it automatically adds both the foreign key and type columns to the table.
```
typo: "creating"

OughtPuts commented on guides/source/association_basics.md:
````diff
@@ -667,7 +872,9 @@ author.books.size
 ```
 
-But what if you want to reload the cache, because data might have been changed by some other part of the application? Just call `reload` on the association:
+NOTE: When we use `author.books`, the data is not immediately loaded from the database. Instead, it sets up a query that will be executed when you actually try to use the data, for example, by calling methods that require data like each, size, empty?, etc. By calling `author.books.load`, you explicitly trigger the query to load the data from the database immediately. This is useful if you know yo ...
````
Suggested change: 'By calling `author.books.load` before calling other methods which use the data... ' will make this really crystal clear.

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -722,13 +929,14 @@ class AddAuthorToBooks < ActiveRecord::Migration[7.2]
-NOTE: If you wish to [enforce referential integrity at the database level][foreign_keys], add the `foreign_key: true` option to the ‘reference’ column declarations above.
-
-[foreign_keys]: active_record_migrations.html#foreign-keys
+NOTE: If you wish to [enforce [referential
```
Formatting is strange here - seems like we might have nested square brackets/nested links?

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -722,13 +929,14 @@ class AddAuthorToBooks < ActiveRecord::Migration[7.2]
+NOTE: If you wish to [enforce [referential
+integrity](https://en.wikipedia.org/wiki/Referential_integrity) at the database
+level](active_record_migrations.html#foreign-keys), add the `foreign_key: true` option to the ‘reference’
+column declarations above.
```
This is also sort of a duplicate of the earlier point about ensuring referential integrity. Do we need it twice?

Ridhwana replied:
@OughtPuts can you point me to the duplicate point please, I think I may be missing where else we mention `foreign_key: true` for referential integrity?

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -1261,11 +1473,7 @@ If you set the `:dependent` option to:
   constraints in your database. The foreign key constraint actions will occur inside the same
```
Worth putting a link where readers can understand the differences between `:destroy` and `:delete`?

Ridhwana replied:
Thanks, makes sense! Instead of a link, I added a little more detail to the section to explain the difference between `delete` and `destroy`

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -1460,7 +1690,7 @@ The `has_one` association creates a one-to-one match with another model. In data
 
 #### Methods Added by `has_one`
```
Worth linking back to where we mentioned this above/or to more about has_one, if we're repeating a bit of information about has_one vs belongs_to here?

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -1518,63 +1757,48 @@ The `association=` method assigns an associated object to this object. Behind th
-##### `:autosave`
-
-If you set the `:autosave` option to `true`, Rails will save any loaded association members and destroy members that are marked for destruction whenever you save the parent object. Setting `:autosave` to `false` is not the same as not setting the `:autosave` option. If the `:autosave` option is not present, then new associated objects will be saved, but updated associated objects will not be save ...
+The [`has`][] association supports numerous options which you can read more about in the [`Options` section of the ActiveRecord Associations API](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_one). We'll discuss some of the common use cases below.
```
Should [`has`] be formatted like this?

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -1600,12 +1824,7 @@ Controls what happens to the associated object when its owner is destroyed:
-##### `:disable_joins`
-
-Specifies whether joins should be skipped for an association. If set to `true`, two or more queries will be generated. Note that in some cases, if order or limit is applied, it will be done in-memory due to database limitations. This option is only applicable on `has_one :through` associations as `has_one` alone does not perform a join.
+because the initial associated object's foreign key will be set to `NULL` value.
```
The sentence above is quite hard to read. Suggest at least 'It's necessary not to set, or leave :nullify option, for those associations...' if not a rewrite?

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -2140,14 +2287,6 @@ end
-
 ##### `:source_type`
 
 The `:source_type` option specifies the source association type for a `has_many :through` association that proceeds through a polymorphic association.
```
Worth linking back to the polymorphic section at this point?

Ridhwana replied:
✅

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -2774,7 +2857,7 @@ If you want to assign an object to a `has_and_belongs_to_many` association witho
 Normal callbacks hook into the life cycle of Active Record objects, allowing you to work with those objects at various points. For example, you can use a `:before_save` callback to cause something to happen just before an object is saved.
 
-Association callbacks are similar to normal callbacks, but they are triggered by events in the life cycle of a collection. There are four available association callbacks:
+Association callbacks are similar to normal callbacks, but they are triggered by events in the life cycle of a collection associated with an Active Record object. There are four available association callbacks:
```
Link here to more context about normal callbacks?

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -2788,16 +2871,22 @@ class Author < ApplicationRecord
 ### Association Extensions
 
-You're not limited to the functionality that Rails automatically builds into association proxy objects. You can also extend these objects through anonymous modules, adding new finders, creators, or other methods. For example:
+Rails allows you to extend the functionality of association proxy objects by adding new finders, creators, or other methods through anonymous modules. This gives you the flexibility to tailor associations to fit your application's needs.
```
I think a beginner may not know what is meant here by 'association proxy objects'... worth trying to rephrase?

OughtPuts reviewed (commented):
Nice one! 🚀 Made a few ease of access suggestions and queried a couple of typos!

p8 commented on guides/source/association_basics.md:
````diff
@@ -1033,15 +1267,24 @@ When you declare a `belongs_to` association, the declaring class automatically g
 ```
 
-Each instance of the `Book` model will have these methods:
+An instance of the `Book` model can have the following methods:
````
Maybe I'm misunderstanding something here,  but doesn't every instance have these methods?
```suggestion
An instance of the `Book` model will have the following methods:
```

p8 commented on guides/source/association_basics.md:
```diff
@@ -1227,25 +1449,15 @@ end
-mentioned methods use the possibly incorrect counter cache column values and always get the results
-from the database, use `counter_cache: { active: false }`. If you also need to specify a custom
-column name, use `counter_cache: { active: false, column: :my_custom_counter }`.
+Using counter caches on existing large tables can be troublesome. To avoid locking the table for too long, the column values must be backfilled separately from the column addition. This backfill must also happen before using `:counter_cache`; otherwise, methods like `size`, `any?`, etc., which rely on counter caches, may return incorrect results.
```
The whole `counter_cache` section might deserve it's own section (maybe under tips and tricks?).
Right now it's not even showing up in the sidebar and this is the only guide that mentions them.

p8 commented on guides/source/association_basics.md:
```diff
@@ -1278,36 +1486,31 @@ class Book < ApplicationRecord
-`user_id` value as the `guid` value of `@user`.
+When you execute `@user.todos.create`, the `@todo` record will have its `user_id` value set to the `guid` value of `@user`.
 
 ##### `:inverse_of`
```
We already discuss `inverse_of` in-depth in the Bi-directional Association section.
Clear examples also require a `has_many` and a `belongs_to`, so I think it would be ok to remove it here?

p8 commented on guides/source/association_basics.md:
```diff
@@ -1319,25 +1522,13 @@ class Book < ApplicationRecord
 
+In this example, `Author` has many `books`, and each `Book` belongs to an `Author`. The `:inverse_of` option ensures that Rails correctly sets up the bidirectional association, allowing efficient and accurate access to related objects. For more details, see the [bi-directional association](#bi-directional-associations) section.
+
 ##### `:optional`
```
This one is mentioned early on in the guide as well. I think it would be ok to remove it.

p8 commented on guides/source/association_basics.md:
```diff
@@ -1319,25 +1522,13 @@ class Book < ApplicationRecord
-
-Enforces strict loading every time the associated record is loaded through this association.
 
 ##### `:touch`
```
The touch option also applies to most other associations.
Maybe we can make that a separate section as well, instead of duplicating it for every association?

p8 reviewed (commented):
This is looking great @Ridhwana!
I only got halfway there, and continue later.
It's a massive guide again.
I seem to have had some similar suggestion as @OughtPuts made 😄

p8 commented on guides/source/association_basics.md:
```diff
@@ -2166,18 +2305,6 @@ class Hardback < ApplicationRecord; end
-
-If you set the `:validate` option to `false`, then new associated objects will not be validated whenever you save this object. By default, this is `true`: new associated objects will be validated when this object is saved.
-
 #### Scopes for `has_many`
```
All these "Scopes" sections for each association have a lot of duplication.
I think it would make sense to have a single top section (maybe "Scopes for associations" or "Customizing Association Scope") and remove the duplication.

These sections also describe common ActiveRecord query methods, I'm not sure they add much value. Maybe the examples and why you would use them add some value?

p8 reviewed (commented):
Finished my review. Loved the rewrite of the STI and Delegated types sections.

p8 commented on guides/source/association_basics.md:
```diff
@@ -1319,25 +1522,13 @@ class Book < ApplicationRecord
 If you set the `:optional` option to `true`, then the presence of the associated
 object won't be validated. By default, this option is set to `false`.
 
-##### `:polymorphic`
-
-Passing `true` to the `:polymorphic` option indicates that this is a polymorphic association. Polymorphic associations were discussed in detail <a href="#polymorphic-associations">earlier in this guide</a>.
-
-
-##### `:required`
-
-When set to `true`, the association will also have its presence validated. This will validate the association itself, not the id. You can use `:inverse_of` to avoid an extra query during validation.
-
-NOTE: required is set to `true` by default and is deprecated. If you don’t want to have association presence validated, use `optional: true`.
```
I think it makes sense to delete this options, as they are handled elsewhere or deprecated.

Ridhwana committed:
update based on feedbcak

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana reviewed (commented):
Thank you @p8 and @OughtPuts. I've implemented some of the feedback, I dropped one follow up question @OughtPuts. I'm busy implementing the rest of the feedback.

Ridhwana committed:
update  based on feedbcak

Ridhwana committed:
update  based on feedbcak

OughtPuts commented:
https://github.com/Ridhwana/rails/pull/8#discussion_r1678411378 

@Ridhwana hmmm I've had a look back and actually I think it's OK to be left as is - although I can't actually remember what I had in mind at the time 😅 any other similar statements that I could find have different enough contexts I think!

p8 commented:
@OughtPuts @Ridhwana I thought the following sentence about "Data integrity" had some overlap with "Referential integrity":

https://github.com/Ridhwana/rails/blob/c84124886010b0e55b6e4dbeb98a13d2ad25dfaa/guides/source/association_basics.md?plain=1#L320-L322

Ridhwana committed:
merge main

Ridhwana committed:
feat: move STI and Delegated Types  a few sections up

Ridhwana committed:
add counter cache to Tips and Tricks

Ridhwana committed:
feat: move the paragraph to comments in the migration

Ridhwana committed:
feat: update the association section to reduce repetition

Ridhwana committed:
feat: update the 'Data integrity' had some overlap with 'Referential integrity'

Ridhwana committed:
add an INFO section to explain macro style calls

Ridhwana committed:
feat: update the diff between delete and destroy

Ridhwana committed:
chore: rename the heading

Ridhwana committed:
chore: update the detailed association references section

Ridhwana commented:
Hi @OughtPuts and @p8,

Thank you so much for the valuable feedback on the PR! I've addressed all your comments and made a couple of important notes for re-review:

- I've moved the "STI" and "Delegated Types" sections after "Advanced Associations."
- The "Counter Cache" section has been relocated to "Tips, Tricks, and Warnings."
- I've restructured the "Detailed Association References" section. I've abstracted the "Scopes" and "Options" sections from the "References" section to avoid duplication across the different types of associations, and I’ve included notes where certain elements may not apply to all associations.

I still have some uncertainty about the necessity of the "References" (association_basics.html#references) section, as it overlaps with the API documentation. However, I believe we provide useful examples that add clarity beyond what is covered in the API. I'm open to any suggestions on how we might improve this section.

I apologize for the slight delay in addressing your feedback. Rewriting this section was quite challenging and it was a huge PR. I wanted to ensure that we find the best approach to enhance clarity and make it more beginner-friendly while also covering some advanced topics.

Thank you again for your support!

OughtPuts commented on guides/source/association_basics.md:
```diff
@@ -594,1804 +784,1332 @@ class Book < ApplicationRecord
+------------------------------
 
-If you want to assign an object to a `has_one` association without saving the object, use the `build_association` method.
+### References
```
Just to double check, did you mean for us to take a look at just this 'References' section, or the entire parent 'Detailed Association Reference' parent?

I think it's a tricky one to get right. Gut feel is that there's a lot of (maybe unnecessary) information here, but I think that's just because it almost serves as a quick glossary of methods that people will refer to. It's also repetitive, and I've wondered about formatting slightly differently to sort that, but again, if you're coming here to find out about the available methods for `has_one`, for example, you probably want them all in one place.

I think it's clear though, and serves a purpose. Maybe let's see what the community come back with when we put the changes out to them - if they think it's unwieldy, then maybe there's a way of splitting this section out to somewhere else.

Ridhwana replied:
> Just to double check, did you mean for us to take a look at just this 'References' section, or the entire parent 'Detailed Association Reference' parent?

Mostly the "References" section although I've now pulled "Scopes" and "options" out into their own sections, so I'm referring to those three sections. 

> Gut feel is that there's a lot of (maybe unnecessary) information here

I feel very similarly and I'm concerned that it's not super useful and like you say repetitive (although we've done a better job with the examples here in the guides compared to the API) 

> Maybe let's see what the community come back with when we put the changes out to them - if they think it's unwieldy, then maybe there's a way of splitting this section out to somewhere else.

This sounds like a sane way forward - lets see if Petrik has any immediate suggestions and if not we'll get community feedback.

p8 replied:
I'm also not sure about the "References" section. It reads more like the API documentation, listing a lot of methods, instead of a guide (this PR improves this a lot already). I'm in favor of removing as much as possible from this section. On the other hand associations are a heavily used Rails API, so it deserves more than just linking to the API documentation.

The descriptions of the added methods are sort of hidden in the "References" section now.
Maybe we can merge the methods for each association into the "Types of Associations" sections?
This makes sure each association is described in a single section.
As the associations are such an important API, maybe each of the six associations can be a top section?
This would also clarify what each section contains in the sidebar.

The methods added for `has_many`, `has_and_belongs_to_many` and `has_many :through` are mostly identical.
So I think in the "Reference" section we only need to mention the methods for `has_many` and mention they are similar for  `has_and_belongs_to_many` and `has_many :through` (this one already seems deleted in the PR).

I hope these suggestions are somewhat helpful. Otherwise, let's get this out to the community.

Ridhwana replied:
@p8 these suggestions were really helpful, however, there are parts that I'm not sure that I'm following. Here's my thoughts:

> Maybe we can merge the methods for each association into the "Types of Associations" sections?
This makes sure each association is described in a single section.

This makes sense to me, and cleans it up nicely!

> As the associations are such an important API, maybe each of the six associations can be a top section?

I think that you're suggesting that we should move this out of the "Types of Associations" section and make them level-1 headers. If this is correct, I'm not sure that I feel aligned with this, as we will only show one subheading "Methods added by .." under each section. I like that we have the heading "Types of Associations" to group them. 

> The methods added for has_many, has_and_belongs_to_many and has_many :through are mostly identical. So I think in the "Reference" section we only need to mention the methods for has_many and mention they are similar for has_and_belongs_to_many and has_many :through (this one already seems deleted in the PR).

I've removed the "References" sections and added the methods for each association into the "Types of Associations" sections. However, the duplication under those sections for the methods is not ideal. However since we're moving them under the associations, it feels strange to extract them. We also don't have API docs linked to `has_one :through` and `has_many :through` which looks odd.

Let me know what you think of the changes. I'll create a community PR  if you feel its good enough for now? and we can continue this discussion there and maybe also get the communities thoughts.

p8 commented on guides/source/association_basics.md:
```diff
@@ -7,20 +7,54 @@ This guide covers the association features of Active Record.
+      t.timestamps
+    end
+  end
+end
```
````suggestion
end
```
````

p8 commented on guides/source/association_basics.md:
````diff
@@ -594,1804 +784,1332 @@ class Book < ApplicationRecord
 ```
 
-##### `includes`
+### Counter Cache
````
Looking at the guide as a whole, I think it makes more sense to maybe move the "Counter Cache" section under the "Detailed Associations References" section.

p8 commented on guides/source/association_basics.md:
```diff
@@ -594,1804 +784,1332 @@ class Book < ApplicationRecord
+NOTE: When initializing a new `has_one` or `belongs_to` association you must use the `build_` prefix to build the association, rather than the `association.build` method that would be used for `has_many` or `has_and_belongs_to_many` associations. To create one, use the `create_` prefix.
 
-##### `collection`
+###### `association`, `reload_association`, and `reset_association`
```
I think grouping related methods makes sense.
The header would benefit from describing the use case.
Maybe:
```suggestion
###### Retreiving the association
```

p8 commented on guides/source/association_basics.md:
```diff
@@ -594,1804 +784,1332 @@ class Book < ApplicationRecord
-WARNING: Additionally, objects will be destroyed if they're associated with `dependent: :destroy`, and deleted if they're associated with `dependent: :delete_all`.
-
-##### `collection.destroy(object, ...)`
+###### `association=(associate)`
```
```suggestion
###### Assigning the Association
```

p8 commented on guides/source/association_basics.md:
```diff
@@ -594,1804 +784,1332 @@ class Book < ApplicationRecord
-The `collection=` method makes the collection contain only the supplied objects, by adding and deleting as appropriate. The changes are persisted to the database.
-
-##### `collection_singular_ids`
+###### `build_association(attributes = {})`, `create_association(attributes = {})` and `create_association!(attributes = {})`
```
I guess `build_association`, `create_association` could still belong under `Assigning the Association`
```suggestion
```

p8 commented on guides/source/association_basics.md:
```diff
@@ -594,1804 +784,1332 @@ class Book < ApplicationRecord
+NOTE: When initializing a new `has_one` or `belongs_to` association you must use the `build_` prefix to build the association, rather than the `association.build` method that would be used for `has_many` or `has_and_belongs_to_many` associations. To create one, use the `create_` prefix.
 
-##### `:class_name`
+###### `association`, `reload_association`, and `reset_association`
```
```suggestion
###### Retreiving the association
```

p8 commented on guides/source/association_basics.md:
```diff
@@ -594,1804 +784,1332 @@ class Book < ApplicationRecord
-* `:nullify` causes the foreign key to be set to `NULL`. Polymorphic type column is also nullified on polymorphic associations. Callbacks are not executed.
-* `:restrict_with_exception` causes an `ActiveRecord::DeleteRestrictionError` exception to be raised if there are any associated records
-* `:restrict_with_error` causes an error to be added to the owner if there are any associated objects
+###### `association=(associate)`
```
```suggestion
###### Assigning the Association
```

p8 commented on guides/source/association_basics.md:
````diff
@@ -594,1804 +784,1332 @@ class Book < ApplicationRecord
+```
 
-Specifies whether joins should be skipped for an association. If set to true, two or more queries will be generated. Note that in some cases, if order or limit is applied, it will be done in-memory due to database limitations. This option is only applicable on `has_many :through associations` as has_many alone do not perform a join.
+###### `build_association(attributes = {})`, `create_association(attributes = {})` and `create_association!(attributes = {})`
````
```suggestion
```

p8 commented on guides/source/association_basics.md:
````diff
@@ -594,1804 +784,1332 @@ class Book < ApplicationRecord
+```
 
-##### `collection.build(attributes = {})`
+##### Checking for Existing Associations
````
I think this makes a lot more sense as a header.

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
move the counter cache section under Detailed Association References

Ridhwana committed:
typo

Ridhwana committed:
chore: move the scopes using associated owner under the scopes section

Ridhwana committed:
chore: move the scopes using associated owner under the scopes section

Ridhwana committed:
move has_many and has_one to the types of asssociations sections

Ridhwana committed:
move has_and_belongs_to_many to the types of asssociations sections

Ridhwana committed:
remove references section

Ridhwana committed:
Merge branch 'main' into Ridhwana/active-record-associations-documentation

Ridhwana committed:
update the header

Ridhwana committed:
chore: rubocop and lint fixes

Ridhwana committed:
column wrap

Ridhwana committed:
chore: rubocop and lint fixes

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Rizwan Reza <github@rizwanreza.com>

Ridhwana committed:
add a distinction between what Rails does and wjat the Database does

Ridhwana committed:
add considerations section to STI that describes some drawbacks

Ridhwana committed:
use references instead of belongs_to

Ridhwana committed:
commit to belongs_to alias

Ridhwana committed:
address feedback

Ridhwana committed:
connection should be relationship

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
merge

Ridhwana committed:
respond to feedback

Ridhwana committed:
updates for composite keys for associations

Ridhwana committed:
update the format

Ridhwana committed:
group the collection methods

Ridhwana committed:
Update guides/source/association_basics.md

Co-authored-by: Petrik de Heus <petrik@deheus.net>

Ridhwana committed:
update column wrap

Ridhwana committed:
format the three sentences that are giving issues
